File: typescript/problem_87.ts
export function validateRules(rules: string[]): boolean {
    const xLo=new Map<string,number>(), xHi=new Map<string,number>();
    const yLo=new Map<string,number>(), yHi=new Map<string,number>();
    const mx=(m:Map<string,number>,k:string,v:number)=>m.set(k,Math.max(m.get(k)??-Infinity,v));
    const mn=(m:Map<string,number>,k:string,v:number)=>m.set(k,Math.min(m.get(k)??Infinity,v));
    for (const rule of rules) {
        const [a,dir,b]=rule.split(' '); const eps=1;
        const yb=yLo.get(b)??0, yt=yHi.get(b)??0, xb=xLo.get(b)??0, xt=xHi.get(b)??0;
        if(dir==='N')  mx(yLo,a,yb+eps);
        else if(dir==='S')  mn(yHi,a,yt-eps);
        else if(dir==='E')  mx(xLo,a,xb+eps);
        else if(dir==='W')  mn(xHi,a,xt-eps);
        else if(dir==='NE'){mx(yLo,a,yb+eps);mx(xLo,a,xb+eps);}
        else if(dir==='NW'){mx(yLo,a,yb+eps);mn(xHi,a,xt-eps);}
        else if(dir==='SE'){mn(yHi,a,yt-eps);mx(xLo,a,xb+eps);}
        else if(dir==='SW'){mn(yHi,a,yt-eps);mn(xHi,a,xt-eps);}
        for(const p of new Set([...xLo.keys(),...xHi.keys()]))
            if((xLo.get(p)??-Infinity)>=(xHi.get(p)??Infinity)) return false;
        for(const p of new Set([...yLo.keys(),...yHi.keys()]))
            if((yLo.get(p)??-Infinity)>=(yHi.get(p)??Infinity)) return false;
    }
    return true;
}

export function problem87(): void {
    console.log("Problem 87:");
    if ( validateRules(["A N B","B NE C","C N A"])) throw new Error("should be invalid");
    if (!validateRules(["A NW B","A N B"]))         throw new Error("should be valid");
    console.log("All Tests Passed");
}

File: go/problem_87.go
package main

import "strings"

// Validate spatial rules using constraint propagation.
// Each rule is "A DIR B" where DIR in {N,NE,E,SE,S,SW,W,NW}.
// We track x (west<0, east>0) and y (south<0, north>0) constraints as inequalities.
func validateRules(rules []string) bool {
	xLo := map[string]float64{}
	xHi := map[string]float64{}
	yLo := map[string]float64{}
	yHi := map[string]float64{}

	for _, rule := range rules {
		parts := strings.Fields(rule)
		if len(parts) != 3 {
			continue
		}
		a, dir, b := parts[0], parts[1], parts[2]
		eps := 1.0
		switch dir {
		case "N":
			yLo[a] = max64(yLo[a], yLo[b]+eps)
		case "S":
			yHi[a] = min64(yHi[a]-eps, yHi[b])
			_ = yHi
		case "E":
			xLo[a] = max64(xLo[a], xLo[b]+eps)
		case "W":
			xHi[a] = min64(xHi[a]-eps, xHi[b])
		case "NE":
			yLo[a] = max64(yLo[a], yLo[b]+eps)
			xLo[a] = max64(xLo[a], xLo[b]+eps)
		case "NW":
			yLo[a] = max64(yLo[a], yLo[b]+eps)
			xHi[a] = min64(xHi[a]-eps, xHi[b])
		case "SE":
			yHi[a] = min64(yHi[a]-eps, yHi[b])
			xLo[a] = max64(xLo[a], xLo[b]+eps)
		case "SW":
			yHi[a] = min64(yHi[a]-eps, yHi[b])
			xHi[a] = min64(xHi[a]-eps, xHi[b])
		}
		for p := range xLo {
			if hi, ok := xHi[p]; ok && xLo[p] >= hi {
				return false
			}
		}
		for p := range yLo {
			if hi, ok := yHi[p]; ok && yLo[p] >= hi {
				return false
			}
		}
	}
	return true
}

func max64(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}
func min64(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

func problem87() {
	println("Problem 87:")
	invalid := []string{"A N B", "B NE C", "C N A"}
	valid := []string{"A NW B", "A N B"}
	if validateRules(invalid) {
		panic("should be invalid")
	}
	if !validateRules(valid) {
		panic("should be valid")
	}
	println("All Tests Passed")
}
